R/derive_param_extreme_record.R

```diff
@@ -0,0 +1,221 @@
+#'
+#'   \item The first or last observation (with respect to the
+#'   `order` variable) for each by group (specified by `by_vars`) from multiple sources
+#'   is selected and added to the input dataset. }
+#'
+#' @return
+#' The input dataset with the first or last observation of each by group
+#' added as new observations.
+#'
+#' @family der_prm_bds_findings
+#' @keywords der_prm_bds_findings
+#'
+#' @export
+#' @examples
+#' aevent_samp <- tibble::tribble(
+#'   ~USUBJID, ~PARAMCD,                       ~PARAM,     ~RSSTDTC,
+#'   "1",          "PD",  "First Progressive Disease", "2022-04-01",
+#'   "2",          "PD",  "First Progressive Disease", "2021-04-01",
+#'   "3",          "PD",  "First Progressive Disease", "2023-04-01"
+#' )
+#'
+#' cm <- tibble::tribble(
+#'   ~STUDYID, ~USUBJID, ~CMDECOD,     ~CMSTDTC,
+#'   "1001",        "1",    "ACT", "2021-12-25"
+#' )
+#'
+#' pr <- tibble::tribble(
+#'   ~STUDYID, ~USUBJID, ~PRDECOD,     ~PRSTDTC,
+#'   "1001",        "1",    "ACS", "2021-12-27",
+#'   "1001",        "2",    "ACS", "2020-12-25",
+#'   "1001",        "3",    "ACS", "2022-12-25",
+#' )
+#' derive_param_extreme_record(
+#'   dataset = aevent_samp,
+#'   sources = list(
+#'     records_source(
+#'       dataset_name = "cm",
+#'       filter = CMDECOD == "ACT",
+#'       new_vars = exprs(
+#'         ADT = convert_dtc_to_dt(CMSTDTC),
+#'         AVALC = CMDECOD
+#'       )
+#'     ),
+#'     records_source(
+#'       dataset_name = "pr",
+#'       filter = PRDECOD == "ACS",
+#'       new_vars = exprs(
+#'         ADT = convert_dtc_to_dt(PRSTDTC),
+#'         AVALC = PRDECOD
+#'       )
+#'     )
+#'   ),
+#'   source_datasets = list(cm = cm, pr = pr),
+#'   by_vars = exprs(USUBJID),
+#'   order = exprs(ADT),
+#'   mode = "first",
+#'   set_values_to = exprs(
+#'     PARAMCD = "FIRSTACT",
+#'     PARAM = "First Anti-Cancer Therapy"
+#'   )
+#' )
+derive_param_extreme_record <- function(dataset = NULL,
+                                        sources,
+                                        source_datasets,
+                                        by_vars = NULL,
+                                        order,
+                                        mode,
+                                        set_values_to) {
+  # Check arguments assertions
+  assert_data_frame(dataset, optional = TRUE)
+  assert_list_of(sources, "records_source")
+  assert_list_of(source_datasets, "data.frame")
+  assert_vars(by_vars, optional = TRUE)
+  assert_character_scalar(
+    mode,
+    values = c("first", "last"),
+    case_sensitive = FALSE
+  )
+  assert_varval_list(set_values_to, accept_expr = TRUE, optional = TRUE)
+
+  source_names <- names(source_datasets)
+
+  # Create Empty list to contain source datasets
+  data_list <- vector("list", length(sources))
+
+  # Evaluate the expressions contained in the sources
+  for (i in seq_along(sources)) {
+    source_dataset <- source_datasets[[sources[[i]]$dataset_name]]
```

Comment:
I was wondering if we could do something like below to remove `source_datasets` argument, so that users don't need to link themselves.
```r
do.call("<-", list("source_dataset", as.name(sources[[1]]$dataset_name)))    
```

Reply:
This works only if the name of the data set coincides with the `dataset_name` field. If `dataset_name = "ae"` but the dataset to be used is `ae_ext` or `myae`, it does not work. We would force the user to rename the dataset to `ae`. This seems not user-friendly to me.

Maybe we could use it as default for the `source_datasets` argument? However, it would somehow hide the dataset. From the function call it would not be obvious that for example the `ae` dataset is used.

Reply:
I was thinking of aligning the dataset name outside of the function call per project level settings (e.g., `myae`, `ae_ext`). 
But the current way makes sense too, I'm fine with aligning it inside the function. Thanks for the discussion!